### ISSUE TITLE
NuGet Package Manager hangs in InfiniteScrollListBoxAutomationPeer.GetChildrenCore

### ISSUE THREAD

**Issue by AArnott** [NuGet/Home#7089]
VS version: 15.8.0 preview 4 (27902.1.d15.8)

I had several documents open in my multiplexingStream solution, including at least one .NET SDK style .csproj file. Then I used the Tools -> Nuget -> Manage solution packages command.
The UI then permanently froze with this:

![image](https://user-images.githubusercontent.com/3548/42249149-801cb462-7ede-11e8-8700-54249a1352fb.png)

I collected [a 32-bit heap dump](https://microsoft-my.sharepoint.com/:u:/p/andarno/Ebe5I-pSnnVOuxzVuitRMkUBj54rhvsX6IgjYKw4uzxKRw?e=7jaWYW). The UI thread callstack is below, which implicates `NuGet_PackageManagement_UI!NuGet.PackageManagement.UI.InfiniteScrollListBoxAutomationPeer.GetChildrenCore`

```
 # ChildEBP RetAddr  
00 0115e474 742ca0d8 win32u!NtUserMsgWaitForMultipleObjectsEx+0xc [windows\core\umode\wow6432\objfre\i386\usrstubs.asm @ 8729] 
01 0115e4e8 742ca003 USER32!RealMsgWaitForMultipleObjectsEx+0x78 [windows\core\ntuser\client\wow6432\objfre\i386\client.cxx @ 1799] 
02 0115e510 54911049 USER32!MsgWaitForMultipleObjectsEx+0x63 [windows\core\ntuser\client\wow6432\objfre\i386\client.cxx @ 1755] 
03 0115e534 772c70a2 vslog!VSResponsiveness::Detours::DetourMsgWaitForMultipleObjectsEx+0x45 [f:\dd\src\vscommon\testtools\vslog\responsetime\vsresponsiveness.cpp @ 810] 
04 0115e56c 772c3cc4 combase!CCliModalLoop::BlockFn+0xfd [onecore\com\combase\dcomrem\callctrl.cxx @ 2377] 
05 0115e624 772d2f7e combase!ClassicSTAThreadWaitForHandles+0xb4 [onecore\com\combase\dcomrem\classicsta.cpp @ 53] 
06 0115e654 549112c3 combase!CoWaitForMultipleHandles+0x8e [onecore\com\combase\dcomrem\sync.cxx @ 123] 
07 0115e678 72c8e27b vslog!VSResponsiveness::Detours::DetourCoWaitForMultipleHandles+0x3e [f:\dd\src\vscommon\testtools\vslog\responsetime\vsresponsiveness.cpp @ 791] 
08 0115e6d0 72c8e220 clr!MsgWaitHelper+0x64 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4256] 
09 (Inline) -------- clr!Thread::DoAppropriateAptStateWait+0x828a2 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4321] 
0a 0115e758 72c0baa4 clr!Thread::DoAppropriateWaitWorker+0x224 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4466] 
0b 0115e7c4 72c8e420 clr!Thread::DoAppropriateWait+0x64 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4133] 
*** WARNING: Unable to verify checksum for C:\WINDOWS\assembly\NativeImages_v4.0.30319_32\WindowsBase\d270b3e5fdab5e7c2cecc5787131f144\WindowsBase.ni.dll
0c 0115ea90 5826b423 clr!SynchronizationContextNative::WaitHelper+0x12c [f:\dd\ndp\clr\src\vm\synchronizationcontextnative.cpp @ 49] 
0d 0115eaa0 71a15cd4 WindowsBase_ni!System.Windows.Threading.DispatcherSynchronizationContext.Wait(IntPtr[], Boolean, Int32)+0x23
0e 0115eab0 72aeeaf6 mscorlib_ni!System.Threading.SynchronizationContext.InvokeWaitMethodHelper(System.Threading.SynchronizationContext, IntPtr[], Boolean, Int32)+0x14
0f 0115eac4 72af1d50 clr!CallDescrWorkerInternal+0x34 [f:\dd\ndp\clr\src\vm\i386\asmhelpers.asm @ 763] 
10 0115eb18 72af7764 clr!CallDescrWorkerWithHandler+0x6b [f:\dd\ndp\clr\src\vm\callhelpers.cpp @ 91] 
11 0115eb88 72c8e1dc clr!MethodDescCallSite::CallTargetWorker+0x16a [f:\dd\ndp\clr\src\vm\callhelpers.cpp @ 655] 
12 (Inline) -------- clr!MethodDescCallSite::Call_RetI4+0xf [f:\dd\ndp\clr\src\vm\callhelpers.h @ 423] 
13 0115ec68 72c09171 clr!Thread::DoSyncContextWait+0xb3 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4889] 
14 0115ecfc 72c0baa4 clr!Thread::DoAppropriateWaitWorker+0x100 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4428] 
15 0115ed68 72c0bc14 clr!Thread::DoAppropriateWait+0x64 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4133] 
16 0115edb4 72b7648b clr!CLREventBase::WaitEx+0x128 [f:\dd\ndp\clr\src\vm\synch.cpp @ 753] 
17 0115edc8 72c80950 clr!CLREventBase::Wait+0x1a [f:\dd\ndp\clr\src\vm\synch.cpp @ 674] 
18 (Inline) -------- clr!Thread::Wait+0x17 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4951] 
19 0115eddc 72c80a8a clr!Thread::Block+0x25 [f:\dd\ndp\clr\src\vm\threads.cpp @ 4908] 
1a 0115ee8c 72c80b6a clr!SyncBlock::Wait+0x189 [f:\dd\ndp\clr\src\vm\syncblk.cpp @ 3564] 
1b (Inline) -------- clr!ObjHeader::Wait+0x24 [f:\dd\ndp\clr\src\vm\syncblk.cpp @ 2743] 
1c (Inline) -------- clr!Object::Wait+0x24 [f:\dd\ndp\clr\src\vm\object.h @ 541] 
1d 0115ef30 71a15f77 clr!ObjectNative::WaitTimeout+0xcb [f:\dd\ndp\clr\src\classlibnative\bcltype\objectnative.cpp @ 315] 
1e 0115ef40 71a27657 mscorlib_ni!System.Threading.Monitor.Wait(System.Object, Int32, Boolean)+0x17
1f 0115ef58 71a275a0 mscorlib_ni!System.Threading.SemaphoreSlim.WaitUntilCountOrTimeout(Int32, UInt32, System.Threading.CancellationToken)+0x4b
20 0115efbc 387113a7 mscorlib_ni!System.Threading.SemaphoreSlim.Wait(Int32, System.Threading.CancellationToken)+0x1f0
21 0115eff0 06d393cb NuGet_PackageManagement_UI!NuGet.PackageManagement.UI.InfiniteScrollListBoxAutomationPeer.GetChildrenCore()+0x47 [E:\A\_work\28\s\src\NuGet.Clients\NuGet.PackageManagement.UI\Xamls\InfiniteScrollListBoxAutomationPeer.cs @ 21] 
*** WARNING: Unable to verify checksum for C:\WINDOWS\assembly\NativeImages_v4.0.30319_32\PresentationCore\df4430399ed08dc8cfc1a9e8a32d2928\PresentationCore.ni.dll
22 0115f008 06d39cab PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.EnsureChildren()+0x23
23 0115f074 07f2e1e7 PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateChildrenInternal(Int32)+0x3b
*** WARNING: Unable to verify checksum for C:\WINDOWS\assembly\NativeImages_v4.0.30319_32\Presentatio5ae0f00f#\21021e7a94b81fee2cd3d17a2cf11808\PresentationFramework.ni.dll
24 0115f080 06d3a29c PresentationFramework_ni!System.Windows.Automation.Peers.ItemsControlAutomationPeer.UpdateChildren()+0x13
25 0115f0c4 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x23c
26 0115f108 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
27 0115f14c 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
28 0115f190 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
29 0115f1d4 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
2a 0115f218 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
2b 0115f25c 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
2c 0115f2a0 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
2d 0115f2e4 06d3a2bb PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
2e 0115f328 068dec42 PresentationCore_ni!System.Windows.Automation.Peers.AutomationPeer.UpdateSubtree()+0x25b
2f 0115f358 068ddf62 PresentationCore_ni!System.Windows.ContextLayoutManager.fireAutomationEvents()+0x122
30 0115f438 068e4d7e PresentationCore_ni!System.Windows.ContextLayoutManager.UpdateLayout()+0x812
31 0115f45c 068e4d55 PresentationCore_ni!System.Windows.ContextLayoutManager.UpdateLayoutCallback(System.Object)+0x1e
32 0115f45c 068e4ce0 PresentationCore_ni!System.Windows.Media.MediaContext+InvokeOnRenderCallback.DoWork()+0x15
33 0115f45c 068e4664 PresentationCore_ni!System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()+0x60
34 0115f498 068e45be PresentationCore_ni!System.Windows.Media.MediaContext.RenderMessageHandlerCore(System.Object)+0x84
35 0115f4b0 5826c062 PresentationCore_ni!System.Windows.Media.MediaContext.RenderMessageHandler(System.Object)+0x6e
36 0115f4c8 5826bfa4 WindowsBase_ni!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)+0x52
37 0115f504 5826e7a2 WindowsBase_ni!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)+0x34
38 0115f54c 5826e6cc WindowsBase_ni!System.Windows.Threading.DispatcherOperation.InvokeImpl()+0xd2
39 0115f568 58433597 WindowsBase_ni!System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)+0x3c
3a 0115f568 71a42925 WindowsBase_ni!MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)+0x3f
3b 0115f5cc 71a42836 mscorlib_ni!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+0xe5
3c 0115f5e0 71a427f1 mscorlib_ni!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+0x16
3d 0115f5fc 584fdd27 mscorlib_ni!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)+0x41
3e 0115f62c 5826e38e WindowsBase_ni!MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)+0x28f8a7
3f 0115f664 5826a9d5 WindowsBase_ni!System.Windows.Threading.DispatcherOperation.Invoke()+0x4e
40 0115f6a4 5826abca WindowsBase_ni!System.Windows.Threading.Dispatcher.ProcessQueue()+0x165
41 0115f6e0 5826c28b WindowsBase_ni!System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)+0x5a
42 0115f71c 5826c1db WindowsBase_ni!MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)+0x9b
43 0115f73c 5826c062 WindowsBase_ni!MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)+0x6b
44 0115f754 5826bfa4 WindowsBase_ni!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)+0x52
45 0115f790 5826a581 WindowsBase_ni!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)+0x34
46 0115f7e8 5826baae WindowsBase_ni!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)+0x131
47 0115f830 03add31a WindowsBase_ni!MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)+0xee
48 0115f864 742cbe6b CLRStub[StubLinkStub]@156164ea03add31a
49 0115f890 742c833a USER32!_InternalCallWinProc+0x2b [internal\onecoreuapwindows\private\lib\i386\Core\ntuser\client\callproc.asm @ 116] 
4a (Inline) -------- USER32!InternalCallWinProc+0x1a [internal\onecoreuapwindows\private\inc\core\ntuser\client\callproc.h @ 87] 
4b 0115f978 742c7bee USER32!UserCallWinProcCheckWow+0x3aa [windows\core\ntuser\client\clmsg.cxx @ 279] 
4c 0115f9f4 742c79d0 USER32!DispatchMessageWorker+0x20e [windows\core\ntuser\client\clmsg.cxx @ 3137] 
4d 0115fa00 0faee3ec USER32!DispatchMessageW+0x10 [internal\onecoreuapwindows\private\inc\core\ntuser\client\cltxt.h @ 1072] 
4e 0115fa20 0faee322 msenv!MainMessageLoop::ProcessMessage+0xef [f:\dd\src\env\msenv\core\main.cpp @ 2012] 
4f 0115fa58 0fb64036 msenv!CMsoCMHandler::EnvironmentMsgLoop+0xde [f:\dd\src\env\msenv\core\msocm.cpp @ 458] 
50 0115fa80 0fb63f83 msenv!CMsoCMHandler::FPushMessageLoop+0x105 [f:\dd\src\env\msenv\core\msocm.cpp @ 366] 
51 0115fac0 0fb63ebe msenv!SCM::FPushMessageLoop+0xc8 [f:\dd\src\env\msenv\mso\core\cistdmgr.cpp @ 2284] 
52 0115fae0 0fb63e85 msenv!SCM_MsoCompMgr::FPushMessageLoop+0x2a [f:\dd\src\env\msenv\mso\core\cistdmgr.cpp @ 3020] 
53 0115fb0c 0fb6202e msenv!CMsoComponent::PushMsgLoop+0x2e [f:\dd\src\env\msenv\core\msocm.cpp @ 713] 
54 0115fb9c 0fbb5e19 msenv!VStudioMainLogged+0x55a [f:\dd\src\env\msenv\core\main.cpp @ 1455] 
55 0115fbc4 0017e308 msenv!VStudioMain+0x7c [f:\dd\src\env\msenv\core\main.cpp @ 1742] 
56 0115fbe0 00170d64 devenv!util_CallVsMain+0x44 [f:\dd\src\appid\lib\utils.cpp @ 1142] 
57 0115fee0 0017e06a devenv!CDevEnvAppId::Run+0x1286 [f:\dd\src\appid\devenv\stub\devenv.cpp @ 978] 
58 0115ff08 0017e183 devenv!WinMain+0xbd [f:\dd\src\appid\devenv\stub\winmain.cpp @ 77] 
59 (Inline) -------- devenv!invoke_main+0x1a [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 94] 
5a 0115ff54 75978484 devenv!__scrt_common_main_seh+0xf6 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 259] 
5b 0115ff68 77682fea KERNEL32!BaseThreadInitThunk+0x24 [base\win32\client\thread.c @ 64] 
5c 0115ffb0 77682fba ntdll!RtlValidSecurityDescriptor+0x11a
5d 0115ffbc 00000000 ntdll!RtlValidSecurityDescriptor+0xea
```

**Comment by jainaashish** [comment]
is it easy to reproduce? This is the consequence of commit# https://github.com/NuGet/NuGet.Client/commit/535b12086fa533d36bbac721bd25a10ed3c276cc Will look into the dump to see why list ItemsLock was not released.

**Comment by jainaashish** [comment]
@AArnott can you share the repro steps for this issue? I've a solution but want to confirm that it fixes it completely.

**Comment by AArnott** [comment]
No, I don't have a consistent repro for this. I only saw the hang once.

**Comment by jainaashish** [comment]
Corresponding VSTS bug which is linked with watson - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/649033

**Comment by jainaashish** [comment]
Merged in dev, but will keep it open until it's merged into release-4.8.0-rtm as well.

**Comment by rrelyea** [comment]
@jainaashish - Please merge into release-4.8.0-rtm this morning.

**Comment by jainaashish** [comment]
Merged into release-4.8.0-rtm branch
